File: exercises/councilling.cpp
/**
 * @author Oskar Arensmeier
 * @date 2025-03-05
 */

#include <iostream>
#include <vector>
#include <set>
#include <string>
#include <map>
#include <climits>
#include <queue>

#define SZ 5000

using namespace std;

map<string, int> names;
map<string, int> parties;
map<string, int> clubs;

vector<int> graph[5000]; // u -> v
int flow[5000][5000];

int bfs(int s, int t, vector<int>& parent) {
    fill(parent.begin(), parent.end(), -1);
    parent[s] = -2;

    queue<pair<int, int>> q;
    q.push({s, INT_MAX});

    while (!q.empty()) {
        int curr      = q.front().first;
        int curr_flow = q.front().second;
        q.pop();

        for (auto next : graph[curr]) {

            if (parent[next] == -1 && flow[curr][next]) {

                parent[next] = curr;
                int new_flow = min(curr_flow, flow[curr][next]);

                if (next == t) {
                    return new_flow;
                }

                q.push({next, new_flow});
            }
        }
    }

    return 0;
}

int max_flow(int s, int t, int n) {
    int curr_flow = 0;
    int new_flow;

    vector<int> parent(n);

    while ((new_flow = bfs(s, t, parent))) {
        curr_flow += new_flow;

        int curr = t;
        while (curr != s) {
            int prev = parent[curr];
            flow[prev][curr] -= new_flow;
            flow[curr][prev] += new_flow;

            curr = prev;
        }
    }

    return curr_flow;
}

int parse()
{
    int N;
    cin >> N;

    int club_count = 0;
    int id = 2;
    for (int i = 0; i < N; ++i) {
        string name, party;
        cin >> name >> party;

        if (parties.find(party) == parties.end()) {
            // source -> party
            graph[0].push_back(id);
            graph[id].push_back(0);

            parties[party] = id;
            id ++;
        }
        
        if (names.find(name) == names.end()) {
            names[name] = id;
            id ++;
        }

        int n = names[name];
        int p = parties[party];

        // party -> name
        graph[p].push_back(n);
        graph[n].push_back(p);
        
        int C;
        cin >> C;
        for (int j = 0; j < C; ++j) {
            string club;
            cin >> club;

            if (clubs.find(club) == clubs.end()) {
                // club -> sink
                graph[id].push_back(1);
                graph[1].push_back(id);

                clubs[club] = id;
                id ++;
            }

            int c = clubs[club];

            // name -> club
            graph[n].push_back(c);
            graph[c].push_back(n);
        }
    }

    int max_party_flow = (clubs.size() - 1) / 2;

    for (int p : graph[0]) {
        // source -> party
        flow[0][p] = max_party_flow;
    }

    for (auto & party: parties) {
        int p = party.second;
        for (auto n : graph[p]) {
            // party -> name
            flow[p][n] = 1;
        }
    }

    for (auto & name : names) {
        int n = name.second;
        for (auto c : graph[n]) {
            // name -> club
            flow[n][c] = 1;
        }
    }

    for (auto & club : clubs) {
        int c = club.second;
        // club -> sink
        flow[c][1] = 1;
    }

    return id;
}

int main()
{
    cin.sync_with_stdio(false);
    cin.tie(nullptr);

    int T;
    cin >> T;

    for (int t = 0; t < T; ++t) {
        names.clear();
        parties.clear();
        clubs.clear();

        fill(&flow[0][0], &flow[0][0] + SZ * SZ, 0);
        for (int i = 0; i < SZ; ++i) {
            graph[i].clear();
        }

        int count = parse();

        /* for (int i = 0; i < count; ++i) {
            for (int j = 0; j < count; ++j) {
                cout << flow[i][j] << ' ';
            }
            cout << '\n';
        }
        cout << flush; */

        /* for (int u = 0; u < count; ++u) {
            cout << u << ": ";
            for (int v : graph[u]) {
                cout << v << ' ';
            }
            cout << '\n';
        }
        cout << flush; */

        int f = max_flow(0, 1, count);

        if (f != clubs.size()) {
            cout << "Impossible.\n";
        } else {
            for (auto & n: names) {
                for (auto & c: clubs) {
                    if (flow[c.second][n.second] == 1 && flow[n.second][c.second] == 0) {
                        cout << n.first << ' ' << c.first << '\n';
                    }
                }
            }
        }

        cout << '\n';
    }

    cout << flush;
}

File: exercises/councilling.py
"""
@author Oskar Arensmeier
@date 2024-02-27
"""
import sys

from copy import deepcopy

adj_list = list[list[int]]

def push(
    flow: adj_list,
    capacity: adj_list,
    excess: list[int],
    u: int,
    v: int,
) -> None:
    d = min(excess[u], capacity[u][v] - flow[u][v])
    flow[u][v] += d
    flow[v][u] -= d
    excess[u] -= d
    excess[v] += d

def relabel(
    N: int,
    flow: adj_list,
    capacity: adj_list,
    excess: list[int],
    height: list[int],
    u: int
) -> None:
    d = float("inf")

    for v in range(N):
        if capacity[u][v] - flow[u][v] > 0:
            d = min(d, height[v])

    if d < float("inf"):
        height[u] = d + 1

def fin_max_height(
    N: int,
    excess: list[int],
    height: list[int],
    s: int,
    t: int
) -> list[int]:
    max_height_vertices = []

    for v in range(N):
        if v != s and v != t and excess[v] > 0:

            if max_height_vertices and height[v] > height[max_height_vertices[0]]:
                max_height_vertices = [] # clear

            if not max_height_vertices or height[v] == height[max_height_vertices[0]]:
                max_height_vertices.append(v)

    return max_height_vertices

def max_flow(
    N: int,
    flow: adj_list,
    capacity: adj_list,
    excess: list[int],
    height: list[int],
    s: int,
    t: int,
):

    for v in range(N):
        if s != v:
            push(flow, capacity, excess, s, v)

    current = []
    while True:
        current = fin_max_height(N, excess, height, s, t)

        if not current:
            break

        for u in current:
            pushed = False

            for v in range(N):

                if not excess[u]:
                    break

                if capacity[u][v] - flow[u][v] > 0 and height[u] == height[v] + 1:
                    push(flow, capacity, excess, u, v)
                    pushed = True

            if not pushed:
                relabel(N, flow, capacity, excess, height, u)

    return excess[t]

def main():
    n = int(sys.stdin.readline().strip())

    names       = list()
    all_parties = set()
    all_clubs   = set()

    lines = []
    for _ in range(n):
        name, party, _, *clubs = sys.stdin.readline().strip().split()
        #print(name, party, clubs)
        lines.append((name, party, clubs))

        names.append(name)
        all_parties.add(party)
        all_clubs.update(clubs)

    # +2 for start and end nodes
    N = len(names) + len(all_parties) + len(all_clubs) + 2

    id_map = dict()
    id_map["start"] = 0
    id_map["end"]   = 1

    for i, name in enumerate(names):
        id_map[name] = i + 2
    for i, party in enumerate(all_parties):
        id_map[party] = i + 2 + len(names)
    for i, club in enumerate(all_clubs):
        id_map[club] = i + 2 + len(names) + len(all_parties)

    # Push-relabel algorithm
    #graph    = [[0 for _ in range(N)] for _ in range(N)]
    flow     = [[0 for _ in range(N)] for _ in range(N)]
    capacity = [[0 for _ in range(N)] for _ in range(N)]
    excess   = [0 for _ in range(N)]
    height   = [0 for _ in range(N)]

    height[id_map["start"]] = N
    excess[id_map["start"]] = N

    party_max_capacity = len(names) // 2 - 1

    for name, party, clubs in lines:
        #graph[id_map["start"]][id_map[party]] = 1
        #graph[id_map[party]][id_map[name]]    = 1

        capacity[id_map["start"]][id_map[party]] = max(0, party_max_capacity)
        capacity[id_map[party]][id_map[name]]    = 1

        for club in clubs:
            #graph[id_map[name]][id_map[club]] = 1
            #graph[id_map[club]][id_map["end"]] = 1

            capacity[id_map[name]][id_map[club]] = 1
            capacity[id_map[club]][id_map["end"]] = 1

    result = max_flow(N, flow, capacity, excess, height, id_map["start"], id_map["end"])

    if result != len(all_clubs):
        print("Impossible.")
        return

    for name, party, clubs in lines:
        for club in clubs:
            if flow[id_map[name]][id_map[club]] == 1:
                print(name, club)

    print()

if __name__ == "__main__":
    T = int(sys.stdin.readline().strip())
    for _ in range(T):
        main()